moshi/src/main/java/com/squareup/moshi/JsonReader.java

```diff
@@ -569,6 +572,27 @@ public final String getPath() {
     return JsonScope.getPath(stackSize, scopes, pathNames, pathIndices);
   }
 
+  /** Returns the tag value for the given class key. */
+  @SuppressWarnings("unchecked")
+  @CheckReturnValue
+  public final @Nullable <T> T tag(Class<T> clazz) {
+    if (tags == null) {
+      return null;
+    }
+    return (T) tags.get(clazz);
+  }
+
+  /** Assigns the tag value using the given class key and value. */
+  public final <T> void setTag(Class<T> clazz, T value) {
+    if (clazz != value.getClass()) {
+      throw new IllegalArgumentException("Tag value must be of type " + clazz.getName());
+    }
+    if (tags == null) {
+      tags = new HashMap<>();
```

Comment:
LinkedHashMap please

https://publicobject.com/2016/02/08/linkedhashmap-is-always-better-than-hashmap/

Reply:
done